hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/AbstractSerializationService.java

```diff
@@ -266,8 +313,12 @@ public final void writeObject(final ObjectDataOutput out, final Object obj) {
         if (obj instanceof Data) {
             throw new HazelcastSerializationException("Cannot write a Data instance, use writeData() instead");
         }
+        // We don't expect obj being compact serializable is a common usage.
```

Comment:
writeObject is called from ObjectDataOutput which is on the DataSerializable API. 
Also you can access to it from Portable via PortableWriter.getRawDataOutput. If someone happens to write a compactSerializable there, it could lead to a problem.

May be instead we can throw an exception if serializer is compact here to avoid interesting hard to resolve problems. What do you think?

Reply:
I've just realized that there is another usage. 
The serializers that we have implemented for java collections use writeObjects.
See `AbstractCollectionStreamSerializer` as an example. 

So if user wants to put an List<Employee> where employee will be Compact Serialized this method will be used. 

Then the correct thing to do would be set includeSchema to false as all user API's will do so. 
There is a risk that we internally use the same API in wrong places like `HotRestart` or `Wan`. But I don't see any precaution for it without causing bad performance on common usages.